### PR TITLE
[7.x] try to fix phpstan & phpstorm reported errors (v1)

### PR DESCRIPTION
**Pull request by voku**
phpstan / phpstorm are reporting thousands of errors and it's hard to analyze the code for them, because of some magic in the code. But I think we can fix many of them without more magic, but with some assert() calls, type hints and better phpdocs.

- remove unnecessary "array_values()" calls (reported by Php Inspections (EA Ultimate))
- fix many phpdocs (reported by phpstorm)
- add some generics annotations (more work needed - see https://suckup.de/2020/02/modern-phpdoc-annotations/)
    -> this is very helpfull for methods / functions that have "mixed" types
    -> support for phpstan + pslam both via "@psalm-*" 
- add one more test for "Arr"

... I don't know if these code "improvements" are even wanted, please give feedback, so that I know where to focus on: auto-completion fixes for the IDE or phpstan support or phpdoc fixes ...

---
This change is [Reviewable](https://reviewable.io/reviews/laravel/framework/31486)

### CODE REVIEW

File: src/Illuminate/Database/Connectors/MySqlConnector.php [hunk 1]
```diff
@@ -44,7 +44,7 @@ public function connect(array $config)
      *
      * @param  \PDO  $connection
      * @param  array  $config
-     * @return void
+     * @return void|\PDO
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Connectors/MySqlConnector.php]: 👎

File: src/Illuminate/Database/Eloquent/Collection.php [hunk 2]
```diff
@@ -10,6 +10,11 @@
 use Illuminate\Support\Str;
 use LogicException;
 
+/**
+ * @template TKey of array-key
+ * @template T
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Collection.php]: Can we avoid using templates in the core?

File: src/Illuminate/Support/Collection.php [hunk 3]
```diff
@@ -24,6 +31,8 @@ class Collection implements ArrayAccess, Enumerable
      *
      * @param  mixed  $items
      * @return void
+     *
+     * @psalm-param array<TKey,T>|mixed $items
```
Comment (GrahamCampbell) [inline src/Illuminate/Support/Collection.php]: 👎

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]
This is unlikely to be merged with all the extra doc.

**Comment by taylorotwell** [comment]
Thanks for looking into this. I think for now I'm mainly only interested in fixing things to help PHPStorm. I don't particularly care about PHPStan as to be honest Laravel does embrace the dynamic side of PHP and I enjoy and embrace that and am not particularly interested in needing to satisfy tools like PHPStan, etc. Maybe in the future though!

**Comment by ondrejmirtes** [comment]
> I think for now I'm mainly only interested in fixing things to help PHPStorm. I don't particularly care about PHPStan

PhpStorm now understands PHPStan's generic syntax for better understanding of collections so this decision might be worthy reconsidering.

**Comment by driesvints** [comment]
The problem with this PR mainly is probably that it's too large. If it could be sent in as smaller incremental fixes then we could reconsider those.

Things like the `psalm-param` annotations won't happen though as we don't want to optimise for specific tools here (sorry). And there's a lot of unrelated things in this PR like code style changes and FQN removals in DocBlocks.

**Comment by GrahamCampbell** [comment]
People can already provide stubs for phpstan and psalm to read, anyway, without the framework source needing to change, so if a third party wants to maintain some stubs, they are more than welcome to.

File: src/Illuminate/Database/Eloquent/Collection.php (hunk 2 above)
Comment (brandonkelly) [reply src/Illuminate/Database/Eloquent/Collection.php:15]: `@template T` is exactly how PhpStorm does it in their example here: https://blog.jetbrains.com/phpstorm/2021/07/phpstorm-2021-2-release/#collections_with_template

**Pull request closed without merging**